packages/go_router/lib/src/route.dart

```diff
@@ -279,3 +317,121 @@ class GoRoute {
+/// passed to the /b/details route so that it displays on the root Navigator
+/// instead of the ShellRoute's Navigator:
+///
+/// ```
+/// final GlobalKey<NavigatorState> _rootNavigatorKey =
+//     GlobalKey<NavigatorState>();
+///
+///   final GoRouter _router = GoRouter(
+///     navigatorKey: _rootNavigatorKey,
+///     initialLocation: '/a',
+///     routes: [
+///       ShellRoute(
+///         navigatorKey: _shellNavigatorKey,
+///         builder: (context, state, child) {
+///           return ScaffoldWithNavBar(child: child);
+///         },
+///         routes: [
+///           // This screen is displayed on the ShellRoute's Navigator.
+///           GoRoute(
+///             path: '/a',
+///             builder: (context, state) {
+///               return const ScreenA();
+///             },
+///             routes: <RouteBase>[
+///               // This screen is displayed on the ShellRoute's Navigator.
+///               GoRoute(
+///                 path: 'details',
+///                 builder: (BuildContext context, GoRouterState state) {
+///                   return const DetailsScreen(label: 'A');
+///                 },
+///               ),
+///             ],
+///           ),
+///           // Displayed ShellRoute's Navigator.
+///           GoRoute(
+///             path: '/b',
+///             builder: (BuildContext context, GoRouterState state) {
+///               return const ScreenB();
+///             },
+///             routes: <RouteBase>[
+///               // Displayed on the root Navigator by specifying the
+///               // [parentNavigatorKey].
+///               GoRoute(
+///                 path: 'details',
+///                 parentNavigatorKey: _rootNavigatorKey,
+///                 builder: (BuildContext context, GoRouterState state) {
+///                   return const DetailsScreen(label: 'B');
+///                 },
+///               ),
+///             ],
+///           ),
+///         ],
+///       ),
+///     ],
+///   );
+/// ```
+///
+/// The widget built by the matching sub-route becomes the child parameter
+/// of the [builder].
+///
+/// For example:
+///
+/// ```
+/// ShellRoute(
+///   path: '/',
+///   builder: (BuildContext context, GoRouterState state, Widget child) {
+///     return Scaffold(
+///       appBar: AppBar(
+///         title: Text('App Shell')
+///       ),
+///       body: Center(
+///         child: child,
+///       ),
+///     );
+///   },
+///   routes: [
+///     GoRoute(
+///       path: 'a'
+///       builder: (BuildContext context, GoRouterState state) {
+///         return Text('Child Route "/a"');
+///       }
+///     ),
+///   ],
+/// ),
+/// ```
+///
+class ShellRoute extends RouteBase {
+  /// Constructs a [ShellRoute].
+  ShellRoute({
+    this.builder,
+    this.pageBuilder,
+    super.routes,
+    GlobalKey<NavigatorState>? navigatorKey,
+  })  : assert(routes.isNotEmpty),
```

Comment:
This should also check every route need to either have null parentnavkey, or the parentnavkey should point to itself

Reply:
johnpryan/shell-route